### ISSUE TITLE
VertexBufferObjectWithVAO continually rebinds attributes

### ISSUE THREAD

**Issue by kurtzmarc** [libgdx/libgdx#5790]
#### Issue details
VertexBufferObjectWithVAO.bindAttributes() will continually rebind attributes every draw cycle.

#### Reproduction steps/code
Observe the following sequence of events:

1. Set `config.useGL30 = true`. Place a model on the screen. Render.
2. `BaseShader.render()` is called.
3. render() calls `getAttributeLocations(renderable.meshPart.mesh.getVertexAttributes()))`:
```
private final int[] getAttributeLocations (final VertexAttributes attrs) {
	tempArray.clear();
	final int n = attrs.size();
	for (int i = 0; i < n; i++) {
		tempArray.add(attributes.get(attrs.get(i).getKey(), -1));
	}
	tempArray.shrink();
	return tempArray.items;
}
```
3. `getAttributeLocations()` returns `tempArray.items` which is the data backing for an IntArray(). We lose the size() of the array at this point. _So this could return an int[16] for an array that is only supposed to be two ints._
4. `BaseShader.render()` calls `Mesh.bind()` which calls `VertexBufferObjectWithVAO.bind()` which calls `bindAttributes()`.
5. `bindAttributes()` checks `stillValid = locations.length == this.cachedLocations.size` which is always false since the number of resolved attributes ends up being two and the length of the array is 16.

This results in the attributes rebinding every draw cycle. The core issue is at point 3 where it uses tempArray.items without passing the actual size of the array so we get garbage entries.

#### Version of LibGDX and/or relevant dependencies
Tested on 1.9.3 but issue still appears present on master.

#### Possible resolutions
1. Return a new properly sized array in getAttributeLocations(). This would cause a new allocation every draw cycle.
2. We could call `tempArray.shrink()`. If the number of attributes changes this would still cause a new allocation every draw cycle.
3. We could drop the `stillValid = locations.length == this.cachedLocations.size;` check. This would cause stillValid to be properly set. It's still a questionable decision to leave `tempArray.items` as what is being passed down through the chain.
4. We could pass a `size` parameter along with the `int[] positions` parameter to specify the length of the positions array.

**Comment by mgsx-dev** [comment]
@kurtzmarc in step 3, tmpArray.shrink() is called which "resize" the backing array to its size. That means tmpArray.items doesn't contains garbage entries anymore (length == size).

I run **ModelTest** (with GL30 enabled) on master branch in order to confirm my assumption and there is no issue : locations.length is 2, cachedLocations.size is 2 and stillValid is true.

Did i miss something? if so, could you please give more information in order to reproduce the issue?

**Comment by kurtzmarc** [comment]
Sorry - I see now that this was in fact fixed in 05c635bcff60fb3c6a4ec3c4d0749fb04b003e12 for 1.9.6 when that shrink() call was added. I missed that change when spot checking to see if it was fixed since 1.9.3. Thanks for taking a look.
